src/Domain/CoordinateMaps/TimeDependent/Translation.cpp

```diff
@@ -82,56 +113,186 @@ std::array<tt::remove_cvref_wrap_t<T>, Dim> Translation<Dim>::frame_velocity(
+        for (size_t k = 0; k < get_size(radius); k++) {
+          if (get_element(radius, k) > 1.e-13) {
+            result.get(i, j) = (*f_of_r_).first_deriv(get_element(radius, k)) *
+                               gsl::at(function_of_time, i) *
+                               gsl::at(distance_to_center, j) /
+                               get_element(radius, k);
+          } else {
+            result.get(i, j) = (*f_of_r_).second_deriv(get_element(radius, k)) *
+                               gsl::at(function_of_time, i) *
+                               gsl::at(distance_to_center, j);
+          }
+        }
+      }
+      result.get(i, i) += 1.0;
+    }
+    return result;
+  }
+}
+template <size_t Dim>
+template <typename T>
+tnsr::Ij<tt::remove_cvref_wrap_t<T>, Dim, Frame::NoFrame>
+Translation<Dim>::inv_jacobian(
+    const std::array<T, Dim>& source_coords, const double time,
+    const std::unordered_map<
+        std::string, std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>&
+        functions_of_time) const {
+  if (f_of_r_ == nullptr) {
+    return identity<Dim>(dereference_wrapper(source_coords[0]));
+  } else {
+    return determinant_and_inverse(
+               jacobian(source_coords, time, functions_of_time))
+        .second;
+  }
+}
+
+template <size_t Dim>
+template <typename T>
+std::array<tt::remove_cvref_wrap_t<T>, Dim> Translation<Dim>::coord_helper(
+    const std::array<T, Dim>& source_coords, double time,
+    const std::unordered_map<
+        std::string, std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>&
+        functions_of_time,
+    const size_t function_or_deriv_index) const {
   ASSERT(functions_of_time.count(f_of_t_name_) == 1,
          "The function of time '" << f_of_t_name_
                                   << "' is not one of the known functions of "
                                      "time. The known functions of time are: "
                                   << keys_of(functions_of_time));
-  std::array<tt::remove_cvref_wrap_t<T>, Dim> result{};
-  const auto function_of_time_and_deriv =
-      functions_of_time.at(f_of_t_name_)->func_and_deriv(time);
-  const DataVector& velocity = function_of_time_and_deriv[1];
-  ASSERT(velocity.size() == Dim,
+  const auto func_or_deriv_of_time =
+      functions_of_time.at(f_of_t_name_)
+          ->func_and_deriv(time)[function_or_deriv_index];
+  ASSERT(func_or_deriv_of_time.size() == Dim,
          "The dimension of the function of time ("
-             << velocity.size()
+             << func_or_deriv_of_time.size()
              << ") does not match the dimension of the translation map (" << Dim
              << ").");
+  std::array<tt::remove_cvref_wrap_t<T>, Dim> result{};
   for (size_t i = 0; i < Dim; i++) {
-    gsl::at(result, i) = make_with_value<tt::remove_cvref_wrap_t<T>>(
-        dereference_wrapper(gsl::at(source_coords, i)), velocity[i]);
+    gsl::at(result, i) = gsl::at(source_coords, i);
```

Comment:
Use an `if (function_or_deriv_index == 0)` statement here. If it's true, set the result to `source_coords` like you do now. If it's false, set result to 0...

Reply:
I changed a few things because if I don't set the result the source_coords then I can't use the magnitude function cause source_coords have the wrong typing. I tried to avoid any new allocations this way and this way it keeps the += at the end. I didn't know of a good way to properly size the std::array since it could be a datavector